### ISSUE TITLE
Bug: __class_getitem__ not recognised

### ISSUE THREAD

**Issue by samsja** [python/mypy#14111]
**Bug Report**

I am experienced a very weird behavior (which I am now pretty sure is a bug) with `__class_getitem__`. 

**To Reproduce**

```python
from abc import abstractmethod
from typing import Iterable


class A(Iterable):
    @abstractmethod
    def __class_getitem__(cls, item):
        ...


class B(A):
    def f(self, field: str):

        self.__class__[str]

>>> attribute.py:14: error: Value of type "Type[B]" is not indexable  [index]
```

Here I expect that mypy know that` __class__` is bounded by the class B and therefore implement `__class_getitem__`.

The weirdest part is that removing the parameter in `f` remove the problem

```python
from abc import abstractmethod
from typing import Iterable


class A(Iterable):
    @abstractmethod
    def __class_getitem__(cls, item):
        ...


class B(A):
    def f(self):

        self.__class__[str]

>>> Success: no issues found in 1 source file
```

**Expected Behavior**

mypy recognize that __class__ can be indexed 

**Actual Behavior**

mypy said it is no indexable even though B implement `__class_getitem__`

**Your Environment**

- Mypy version used: 0.990
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): plugins = "pydantic.mypy"

- Python version used: 3.9

FYI: a more advance setup in this PR https://github.com/docarray/docarray/pull/803

**Comment by thomkeh** [comment]
Duplicate of #11501 I believe.

**Comment by AlexWaygood** [comment]
> Duplicate of #11501 I believe.

Yup. Thanks!

**Comment by samsja** [comment]
I am not sure it is totally a duplicate. Did you see that if I don't have a parameters in my function mypy does not complain anymore ?
This work
```python
    def f(self):
        self.__class__[str]
```

This does not
```python
    def f(self, field: str):
        self.__class__[str]
```

Why is the behavior changing ?

**Comment by AlexWaygood** [comment]
The fundamental issue — that `__class_getitem__` is an implementation detail of the typing system at runtime, which mypy has no special recognition of at present — is the same.

**Comment by AlexWaygood** [comment]
> Did you see that if I don't have a parameters in my function mypy does not complain anymore ?

Oh, that's because your "function without parameters" is unannotated, meaning mypy just ignores it. It's a [common footgun](https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code) :)

**Comment by samsja** [comment]
Oh I see ! Then yes it is a duplicate. Thanks for the clarification
